### PR TITLE
WIP: Add support for giftwrap packaging

### PR DESCRIPTION
**Pull request by craigtracey**
Adds support for packaging. This moves all of the openstack source
pulls out of ursula and into packages.

### CODE REVIEW

File: roles/openstack-package/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  apt: pkg="{{ package_name|default(openstack_package.package_name) }}" 
+       update_cache=yes cache_valid_time=3600
+
+- name: setup openstack current base path
+  file: dest=/opt/openstack state=directory
+
+- name: test whether current {{ project_name }} symlink exists
+  shell: 'test -h /opt/openstack/current'
+  register: symlink_exists
+  ignore_errors: True
```
Comment (j2sol) [inline roles/openstack-package/tasks/main.yml]: Why not use the stat module here? You don't have to ignore errors, and the stat details will tell you if it is a symlink or if it exists or not.

File: roles/openstack-package/tasks/alternatives.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+---
+- name: setup alternatives for installed packages
+  shell: update-alternatives --install /usr/local/bin/{{ item }} {{ item }} /opt/openstack/{{ project_name }}/bin/{{ item }} 10
```
Comment (j2sol) [inline roles/openstack-package/tasks/alternatives.yml]: This file doesn't seem to be used.

File: roles/common/tasks/main.yml [hunk 3]
```diff
@@ -1,11 +1,11 @@
 ---
 - include: proxy.yml
 
-- include: bbg-repo.yml
-
 - name: update apt index
   apt: update_cache=yes cache_valid_time=3600
 
+- include: repos.yml
+
```
Comment (j2sol) [inline roles/common/tasks/main.yml]: Any reason for doing this after the apt cache update?

**Comment by j2sol** [comment]
I'm concerned about unversioned client grabs from pip. This would lead to inconsistent client versions across deploys of a given ursula version, as the client versions could change with time. I'm also concerned about service and client interactions (such as nova-compute calling glance client, or nova calling neutron client, or glance calling swift, etc..), and the testing done with one service/client combo not matching what will be deployed.

I'm also curious if we'll be able to specify different versions for specific packages, such as a nova package version that's different from a neutron package version. If we have an update for nova, do we have to rebuild all the packages to the newer version number, or will we be able to just bump one and upgrade it across the cluster?

Speaking of upgrades, it looks like there are no handlers for changes in the package, so if we run the role and the package version changes, the service won't be restarted. Is this intended?

Those are my thoughts after reading through this.

File: roles/openstack-package/tasks/main.yml (hunk 1 above)
Comment (craigtracey) [inline roles/openstack-package/tasks/main.yml]: Good point

File: roles/openstack-package/tasks/alternatives.yml (hunk 2 above)
Comment (craigtracey) [inline roles/openstack-package/tasks/alternatives.yml]: Yeah - caught in the crossfire...we remove. Good catch.

File: roles/common/tasks/main.yml (hunk 3 above)
Comment (craigtracey) [inline roles/common/tasks/main.yml]: Another good one.

**Comment by craigtracey** [comment]
I think unversioned pip clients are OK, but I would be willing to support a pip version for each.  They should always be backwards compat and easy enough to fix if they happen not to be.  As for "nova-compute calling glance client" this is a package issue, not a sys.path pip issue.  Each package includes all the bits it needs in a virtualenv, so any client will be local to its virtualenv...those can be pinned strictly in giftwrap.  These pip clients are strictly for admin/ansible cluster interaction.

I do not intend for us to ship packages of different versions.  Building of packages is easy enough (and numbers free enough :) ), that I personally prefer to ship an integrated set of packages.

This does not include restarts because I think this should be handled by the upgrade orchestration that you have built.

**Comment by j2sol** [comment]
I think my hesitation with unified package numbers and rebuilding every time is that if we want to make a change to .. oh lets say glance, and nothing else, we would have to build new (unchanged) packages for all the services. THEN we would either have to restart all the services to ensure the process is running from the installed package, OR we would have to handle the fact that we have processes running from older packages. I assume that at some point we'll have a fact for what version SHOULD be running and a sensu check to validate that the processes running ARE from that version, and alert if they are not.

Does that make sense?

File: roles/openstack-package/tasks/main.yml (hunk 1 above)
Comment (craigtracey) [inline roles/openstack-package/tasks/main.yml]: On second thought, the stat module does not let you test for whether the path is a symlink or not.

Comment (craigtracey) [inline roles/openstack-package/tasks/main.yml]: Now that I have looked to implement it, I am not sure that I prefer the stat method.

Comment (j2sol) [inline roles/openstack-package/tasks/main.yml]: If you don't want to use stat, at least change the ignore_errors to failed_when: false

One leaves ugly red text on the screen making it hard to find the actual failures, and the other does not.

**Pull request closed without merging**